File: app/lib/core/decode/vote_decode.dart
// app/lib/core/decode/vote_decode.dart
//
// Plan B Task B1.3: decodes a PCM-in "vote"-decoded model's per-frame
// argmax predictions (PcmInferenceRunner's List<FrameResult>) into merged
// Chord segments.
//
// Pipeline:
//   1. Map each frame's classId -> label via ModelSpec.labels.
//   2. Majority-filter smoothing over the class-id sequence (odd
//      smoothingKernel, default 5 ~= 0.46s at frame_dur ~= 0.09288s) to
//      remove single-frame flicker. This mirrors the reference
//      majority_filter_indices in
//      reference/ChordMini/src/evaluation/utils/common.py exactly,
//      including its tie rule and its "too-short sequence" no-op:
//        - edge-clamped window of size kernel (kernel forced odd, >=1).
//        - if the sequence is shorter than the kernel, no smoothing is
//          applied at all (matches reference: `values.size < kernel_size`
//          -> return the sequence unchanged).
//        - for each frame, take the most common class id in the centered,
//          edge-clamped window; on ties, keep the frame's own class if it
//          is among the tied max-count candidates, else take the smallest
//          such candidate (reference iterates candidates in ascending
//          label order and takes candidates[0]).
//   3. Merge consecutive equal smoothed labels into one Chord: start =
//      time of the first frame in the run, end = time of the last frame
//      in the run + frame_dur (frame_dur derived from the consecutive
//      frame time delta), confidence = mean confidence over the run.
//   4. 'N' (no-chord) and 'X' (unknown) runs are kept as Chords like any
//      other label -- never dropped -- so the decoded timeline has no
//      gaps.
import '../inference/pcm_runner.dart';
import '../model_registry.dart';
import '../models.dart';

/// Fallback frame duration (seconds) used only when fewer than 2 frames
/// are available, so a delta between consecutive frame times can't be
/// derived. Matches the PCM-in chordnet/btc hop (2048 samples @ 22050 Hz;
/// see pcm_runner.dart's frame accounting).
const fallbackFrameDur = 2048 / 22050;

/// Applies the reference `majority_filter_indices` categorical majority
/// filter to a sequence of class ids.
List<int> _majorityFilter(List<int> values, int kernelSize) {
  var kernel = kernelSize < 1 ? 1 : kernelSize;
  if (kernel % 2 == 0) kernel += 1;
  final n = values.length;
  if (kernel == 1 || n < kernel) {
    return List<int>.from(values);
  }

  final pad = kernel ~/ 2;
  int atPadded(int paddedIdx) {
    // Edge-clamped index into `values` for a virtual array with `pad`
    // edge-replicated entries on each side (np.pad mode='edge').
    final i = paddedIdx - pad;
    if (i < 0) return values[0];
    if (i >= n) return values[n - 1];
    return values[i];
  }

  final filtered = List<int>.filled(n, 0);
  for (var idx = 0; idx < n; idx++) {
    final counts = <int, int>{};
    for (var w = 0; w < kernel; w++) {
      final v = atPadded(idx + w);
      counts[v] = (counts[v] ?? 0) + 1;
    }
    var maxCount = 0;
    for (final c in counts.values) {
      if (c > maxCount) maxCount = c;
    }
    // Ascending label order, matching np.unique's sorted output.
    final candidates = counts.keys.where((k) => counts[k] == maxCount).toList()
      ..sort();
    final center = values[idx];
    filtered[idx] = candidates.contains(center) ? center : candidates.first;
  }
  return filtered;
}

/// Absorbs chord segments shorter than [minDur] seconds into a neighbor so
/// the timeline has no sub-[minDur] "junk" chords (e.g. a 1-frame D7 between
/// D and G). Repeatedly takes the shortest too-short segment and merges it
/// into its longer-duration neighbor (tie -> previous): the neighbor's label
/// wins, its span extends to cover the gap, confidence becomes the
/// duration-weighted mean. Applies uniformly to 'N'/'X' too. The timeline
/// stays gap-free (every span remains covered).
/// ponytail: O(n^2) scan, fine for song-length chord lists.
List<Chord> _mergeShort(List<Chord> chords, double minDur) {
  final out = List<Chord>.of(chords);
  while (out.length > 1) {
    // Find the shortest segment still under the threshold.
    var shortest = -1;
    var shortestDur = minDur;
    for (var i = 0; i < out.length; i++) {
      final d = out[i].end - out[i].start;
      if (d < shortestDur) {
        shortest = i;
        shortestDur = d;
      }
    }
    if (shortest < 0) break; // nothing left under minDur

    // Pick the longer-duration neighbor to absorb into (tie -> previous).
    final prev = shortest > 0 ? shortest - 1 : -1;
    final next = shortest < out.length - 1 ? shortest + 1 : -1;
    int keep;
    if (prev < 0) {
      keep = next;
    } else if (next < 0) {
      keep = prev;
    } else {
      final prevDur = out[prev].end - out[prev].start;
      final nextDur = out[next].end - out[next].start;
      keep = nextDur > prevDur ? next : prev;
    }

    final s = out[shortest];
    final k = out[keep];
    final sDur = s.end - s.start;
    final kDur = k.end - k.start;
    final conf = (k.confidence * kDur + s.confidence * sDur) / (kDur + sDur);
    out[keep] = Chord.fromJson({
      'chord': k.chord,
      'start': k.start < s.start ? k.start : s.start,
      'end': k.end > s.end ? k.end : s.end,
      'confidence': conf,
    });
    out.removeAt(shortest);

    // Absorbing may leave the kept segment adjacent to an equal label; merge.
    final ki = keep > shortest ? keep - 1 : keep;
    for (final j in [ki + 1, ki - 1]) {
      if (j >= 0 && j < out.length && out[j].chord == out[ki].chord) {
        final a = out[j < ki ? j : ki];
        final b = out[j < ki ? ki : j];
        final aDur = a.end - a.start;
        final bDur = b.end - b.start;
        out[j < ki ? j : ki] = Chord.fromJson({
          'chord': a.chord,
          'start': a.start,
          'end': b.end,
          'confidence':
              (a.confidence * aDur + b.confidence * bDur) / (aDur + bDur),
        });
        out.removeAt(j < ki ? ki : j);
        break;
      }
    }
  }
  return out;
}

/// Decodes [frames] (a PCM-in "vote" model's per-frame argmax
/// predictions) into merged [Chord] segments, smoothing single-frame
/// flicker with a majority filter over an odd-sized [smoothingKernel]
/// (default 5), then absorbing segments shorter than [minChordDur] seconds
/// into a neighbor so the timeline has no sub-[minChordDur] junk chords.
List<Chord> voteDecode(
  List<FrameResult> frames,
  ModelSpec spec, {
  int smoothingKernel = 5,
  double minChordDur = 0.3,
}) {
  if (frames.isEmpty) return [];

  final labels = spec.labels;
  if (labels == null) {
    throw ArgumentError('${spec.name} has no labels in its manifest entry');
  }

  final classIds = [for (final f in frames) f.classId];
  final smoothed = _majorityFilter(classIds, smoothingKernel);

  // frame_dur: consecutive frame time delta (0 if only one frame).
  final frameDur =
      frames.length > 1 ? frames[1].time - frames[0].time : fallbackFrameDur;

  final chords = <Chord>[];
  var runStart = 0;
  for (var i = 1; i <= frames.length; i++) {
    final atRunEnd = i == frames.length || smoothed[i] != smoothed[runStart];
    if (atRunEnd) {
      final runFrames = frames.sublist(runStart, i);
      var sumConf = 0.0;
      for (final f in runFrames) {
        sumConf += f.confidence;
      }
      chords.add(Chord.fromJson({
        'chord': labels[smoothed[runStart]],
        'start': runFrames.first.time,
        'end': runFrames.last.time + frameDur,
        'confidence': sumConf / runFrames.length,
      }));
      runStart = i;
    }
  }
  return _mergeShort(chords, minChordDur);
}

File: server/app/infrastructure/analysis/vote_decode.py
from collections import Counter
from .chords import Chord
from .onnx_infer import Frame
from .manifest import ModelSpec

FALLBACK_FRAME_DUR = 2048 / 22050

def majority_filter(values: list[int], kernel_size: int) -> list[int]:
    kernel = 1 if kernel_size < 1 else kernel_size
    if kernel % 2 == 0:
        kernel += 1
    n = len(values)
    if kernel == 1 or n < kernel:
        return list(values)
    pad = kernel // 2
    def at(padded_idx: int) -> int:
        i = padded_idx - pad
        if i < 0: return values[0]
        if i >= n: return values[n - 1]
        return values[i]
    out = [0] * n
    for idx in range(n):
        counts = Counter(at(idx + w) for w in range(kernel))
        max_count = max(counts.values())
        candidates = sorted(k for k, c in counts.items() if c == max_count)
        center = values[idx]
        out[idx] = center if center in candidates else candidates[0]
    return out

def merge_short_chords(chords: list[Chord], min_dur: float) -> list[Chord]:
    out = list(chords)
    while len(out) > 1:
        shortest, shortest_dur = -1, min_dur
        for i, c in enumerate(out):
            d = c.end - c.start
            if d < shortest_dur:
                shortest, shortest_dur = i, d
        if shortest < 0:
            break
        prev = shortest - 1 if shortest > 0 else -1
        nxt = shortest + 1 if shortest < len(out) - 1 else -1
        if prev < 0:
            keep = nxt
        elif nxt < 0:
            keep = prev
        else:
            keep = nxt if (out[nxt].end - out[nxt].start) > (out[prev].end - out[prev].start) else prev
        s, k = out[shortest], out[keep]
        s_dur, k_dur = s.end - s.start, k.end - k.start
        conf = (k.confidence * k_dur + s.confidence * s_dur) / (k_dur + s_dur)
        out[keep] = Chord(k.chord, min(k.start, s.start), max(k.end, s.end), conf)
        del out[shortest]
        ki = keep - 1 if keep > shortest else keep
        for j in (ki + 1, ki - 1):
            if 0 <= j < len(out) and out[j].chord == out[ki].chord:
                lo, hi = (j, ki) if j < ki else (ki, j)
                a, b = out[lo], out[hi]
                a_dur, b_dur = a.end - a.start, b.end - b.start
                out[lo] = Chord(a.chord, a.start, b.end,
                                (a.confidence * a_dur + b.confidence * b_dur) / (a_dur + b_dur))
                del out[hi]
                break
    return out

def vote_decode(frames: list[Frame], spec: ModelSpec,
                smoothing_kernel: int = 5, min_chord_dur: float = 0.5) -> list[Chord]:
    if not frames:
        return []
    labels = spec.labels
    smoothed = majority_filter([f.class_id for f in frames], smoothing_kernel)
    frame_dur = frames[1].time - frames[0].time if len(frames) > 1 else FALLBACK_FRAME_DUR
    chords: list[Chord] = []
    run_start = 0
    for i in range(1, len(frames) + 1):
        at_end = i == len(frames) or smoothed[i] != smoothed[run_start]
        if at_end:
            run = frames[run_start:i]
            conf = sum(f.confidence for f in run) / len(run)
            chords.append(Chord(labels[smoothed[run_start]], run[0].time,
                                run[-1].time + frame_dur, conf))
            run_start = i
    return merge_short_chords(chords, min_chord_dur)
